### PR TITLE
Fix hostport duplicate chain names

### LINKED ISSUE #55771: Some hostports are unreachable
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
Some hostports are unreachable.

**What you expected to happen**:
All hostports are reachable.

**How to reproduce it (as minimally and precisely as possible)**:
Add multiple hostport when creating an deployment, e.g. ports 57119/55429/56833

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by chenchun**
Fixes bad conversion from int32 to string. Without this patch, getHostportChain/hostportChainName generates the same chain names for ports 57119/55429/56833 of the same pod.

closes #55771

```release-note
Fixes bad conversion in host port chain name generating func which leads to some unreachable host ports.
```

### CODE REVIEW

File: pkg/kubelet/network/hostport/hostport_manager_test.go [hunk 1]
```diff
@@ -198,3 +198,16 @@ func TestHostportManager(t *testing.T) {
 		assert.EqualValues(t, true, port.closed)
 	}
 }
+
+func TestGetHostportChain(t *testing.T) {
+	m := make(map[string]int)
```
Comment (MrHohn) [inline pkg/kubelet/network/hostport/hostport_manager_test.go]: nit: what about `map[string]bool`?

File: pkg/kubelet/network/hostport/hostport_manager.go [hunk 2]
```diff
@@ -247,7 +248,7 @@ func (hm *hostportManager) closeHostports(hostportMappings []*PortMapping) error
 // WARNING: Please do not change this function. Otherwise, HostportManager may not be able to
 // identify existing iptables chains.
```
Comment (MrHohn) [inline pkg/kubelet/network/hostport/hostport_manager.go]: Does that mean by changing the chain name, we are orphaning the old chain/rule? Is that acceptable or anyway to workaround?

@kubernetes/sig-network-pr-reviews

Comment (chenchun) [reply pkg/kubelet/network/hostport/hostport_manager.go]: Add a commit to clean up these old chains/rules as well.

File: pkg/kubelet/network/hostport/hostport_manager_test.go [hunk 3]
```diff
@@ -144,21 +144,21 @@ func TestHostportManager(t *testing.T) {
...
-		"-A KUBE-HP-5N7UH5JAXCVP5UJR -m comment --comment \"pod3_ns1 hostport 8443\" -s 10.1.1.4/32 -j KUBE-MARK-MASQ":                    true,
-		"-A KUBE-HP-5N7UH5JAXCVP5UJR -m comment --comment \"pod3_ns1 hostport 8443\" -m tcp -p tcp -j DNAT --to-destination 10.1.1.4:443": true,
+		"-A KUBE-HP-IJHALPHTORMHHPPK -m comment --comment \"pod1_ns1 hostport 8080\" -s 10.1.1.2/32 -j KUBE-MARK-MASQ":                    true,
+		"-A KUBE-HP-IJHALPHTORMHHPPK -m comment --comment \"pod1_ns1 hostport 8080\" -m tcp -p tcp -j DNAT --to-destination 10.1.1.2:80":  true,
+		"-A KUBE-HP-63UPIDJXVRSZGSUZ -m comment --comment \"pod1_ns1 hostport 8081\" -s 10.1.1.2/32 -j KUBE-MARK-MASQ":                    true,
+		"-A KUBE-HP-63UPIDJXVRSZGSUZ -m comment --comment \"pod1_ns1 hostport 8081\" -m udp -p udp -j DNAT --to-destination 10.1.1.2:81":  true,
+		"-A KUBE-HP-WFBOALXEP42XEMJK -m comment --comment \"pod3_ns1 hostport 8443\" -s 10.1.1.4/32 -j KUBE-MARK-MASQ":                    true,
+		"-A KUBE-HP-WFBOALXEP42XEMJK -m comment --comment \"pod3_ns1 hostport 8443\" -m tcp -p tcp -j DNAT --to-destination 10.1.1.4:443": true,
```
Comment (MrHohn) [inline pkg/kubelet/network/hostport/hostport_manager_test.go]: Could we validate the cleanup logic in unit test so that we will have more confidences?

File: pkg/kubelet/network/hostport/hostport_syncer.go [hunk 4]
```diff
@@ -142,7 +143,7 @@ func writeLine(buf *bytes.Buffer, words ...string) {
 // this because IPTables Chain Names must be <= 28 chars long, and the longer
 // they are the harder they are to read.
 func hostportChainName(pm *PortMapping, podFullName string) utiliptables.Chain {
-	hash := sha256.Sum256([]byte(string(pm.HostPort) + string(pm.Protocol) + podFullName))
+	hash := sha256.Sum256([]byte(strconv.Itoa(int(pm.HostPort)) + string(pm.Protocol) + podFullName))
```
Comment (MrHohn) [inline pkg/kubelet/network/hostport/hostport_syncer.go]: For `hostport_syncer`, is the cleanup logic originally backed in? Might to great to verify it in unit test as well.

**Review by MrHohn — COMMENTED** [review]
Thanks, seems to be a viable solution. Some comments about testing.

**Comment by chenchun** [comment]
@MrHohn Addressed all your comments, PTAL

File: pkg/kubelet/network/hostport/fake_iptables.go [hunk 5]
```diff
@@ -274,6 +275,11 @@ func (f *fakeIPTables) restore(restoreTableName utiliptables.Table, data []byte,
 					}
 				}
 				_, _ = f.ensureChain(tableName, chainName)
+				if !strings.Contains(allLines, "-X "+string(chainName)) {
+					if err := f.FlushChain(tableName, chainName); err != nil {
```
Comment (MrHohn) [inline pkg/kubelet/network/hostport/fake_iptables.go]: Sorry I don't quite get the logic here, why should we flush the chain when there is no `-X CHAIN_NAME`? Explanation would be appreciated.

File: pkg/kubelet/network/hostport/hostport_manager.go (hunk 2 above)
Comment (MrHohn) [reply pkg/kubelet/network/hostport/hostport_manager.go]: In fact, thought about this again, I wonder why it would be an issue changing this function. Can't we assume upon node/kubelet upgrade, all iptables rules will not be retained because node has been restarted? Am I suggesting an unnecessary cleanup? Would be great to have guidance from @thockin and @freehan

File: pkg/kubelet/network/hostport/fake_iptables.go (hunk 5 above)
Comment (chenchun) [reply pkg/kubelet/network/hostport/fake_iptables.go]: I think I should update here to just flush user defined chains. 
The --noflush option for iptables-restore doesn't work for user-defined chains, such as TESTCHAIN, only builtin chains. https://unix.stackexchange.com/questions/134687/how-to-combine-iptables-rulesets

I did a test to confirm this. See below example, the POSTROUTING chain didn't get flush, but KUBE-NODEPORT chain did.
```
[root@kubernetes-master vagrant]# iptables -t nat -S
-P PREROUTING ACCEPT
-P INPUT ACCEPT
-P OUTPUT ACCEPT
-P POSTROUTING ACCEPT
-N KUBE-HP-5N7UH5JAXCVP5UJR
-N KUBE-NODEPORT
-A POSTROUTING -p tcp -m comment --comment "pod3_ns1 hostport 8443" -m tcp --dport 8443 -j KUBE-HP-5N7UH5JAXCVP5UJR
-A KUBE-NODEPORT -p tcp -m comment --comment "pod3_ns1 hostport 8443" -m tcp --dport 8443 -j KUBE-HP-5N7UH5JAXCVP5UJR
[root@kubernetes-master vagrant]# cat nat.log 
# Generated by iptables-save v1.4.21 on Mon Nov 13 03:14:59 2017
*nat
:PREROUTING ACCEPT [65:3900]
:INPUT ACCEPT [65:3900]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:KUBE-HP-5N7UH5JAXCVP5UJR - [0:0]
:KUBE-NODEPORT - [0:0]
COMMIT
# Completed on Mon Nov 13 03:14:59 2017
[root@kubernetes-master vagrant]# iptables-restore --table=nat --noflush < nat.log 
[root@kubernetes-master vagrant]# iptables -t nat -S
-P PREROUTING ACCEPT
-P INPUT ACCEPT
-P OUTPUT ACCEPT
-P POSTROUTING ACCEPT
-N KUBE-HP-5N7UH5JAXCVP5UJR
-N KUBE-NODEPORT
-A POSTROUTING -p tcp -m comment --comment "pod3_ns1 hostport 8443" -m tcp --dport 8443 -j KUBE-HP-5N7UH5JAXCVP5UJR
```

And this behavior is needed to cleanup rules in KUBE-NODEPORT in my tests.

File: pkg/kubelet/network/hostport/hostport_manager.go (hunk 2 above)
Comment (chenchun) [reply pkg/kubelet/network/hostport/hostport_manager.go]: > Can't we assume upon node/kubelet upgrade, all iptables rules will not be retained because node has been restarted?

😕 Is this true? I don't think everyone will restart node upon upgrade. Is this the required step of upgrading to the next release or something?

Comment (MrHohn) [reply pkg/kubelet/network/hostport/hostport_manager.go]: At least that is what happens with [gce/upgrade.sh](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/upgrade.sh) :)

I'm not aware of any supported per-system-component upgrade mechanism in k8s yet, not sure if someone already support that...

File: pkg/kubelet/network/hostport/fake_iptables.go (hunk 5 above)
Comment (MrHohn) [reply pkg/kubelet/network/hostport/fake_iptables.go]: Thanks for the investigation, now I get it. Interesting that this behavior is not documented in iptables manual.

Though I just learned that user-defined chains won't always be flushed, unless they are explicitly mentioned in the input of `iptables-restore --noflush`, as what below comment is referring to:
https://github.com/kubernetes/kubernetes/blob/f28140429e0f3b6b4ef1bb7399d1bbd7d644c933/pkg/kubelet/network/hostport/hostport_syncer.go#L263-L267

In short I believe your fix will work as expected, but the implementation of `fakeIPTables.restore()` still seems problematic as it flushes more than it should...

Comment (MrHohn) [reply pkg/kubelet/network/hostport/fake_iptables.go]: Oops I scan too fast, looks like that is exactly what you have implemented. Could you update it to just flush user defined chains? Thanks.

Comment (chenchun) [reply pkg/kubelet/network/hostport/fake_iptables.go]: Done. Also add a test `TestRestoreFlushRules` for `fackIPtables.restore()`.

Comment (MrHohn) [reply pkg/kubelet/network/hostport/fake_iptables.go]: Awesome, thanks!

**Review by MrHohn — APPROVED** [review]
Thanks, appreciate your works!

/lgtm

**Comment by MrHohn** [comment]
This might worth a release note.
/release-note

/assign @thockin 
for approval.

File: pkg/util/iptables/iptables.go [hunk 6]
```diff
@@ -82,6 +82,7 @@ type Table string
 const (
 	TableNAT    Table = "nat"
 	TableFilter Table = "filter"
+	TableMangle Table = "mangle"
```
Comment (thockin) [inline pkg/util/iptables/iptables.go]: Why is this part of the same PR?  Seems unrelated?

File: pkg/kubelet/network/hostport/hostport_manager.go [hunk 7]
```diff
@@ -177,6 +178,8 @@ func (hm *hostportManager) Remove(id string, podPortMapping *PodPortMapping) (er
 	chainsToRemove := []utiliptables.Chain{}
 	for _, pm := range hostportMappings {
 		chainsToRemove = append(chainsToRemove, getHostportChain(id, pm))
+		// TODO remove this, please refer https://github.com/kubernetes/kubernetes/pull/55153
+		chainsToRemove = append(chainsToRemove, getBugyHostportChain(id, pm))
```
Comment (thockin) [inline pkg/kubelet/network/hostport/hostport_manager.go]: s/Bugy/Buggy/

File: pkg/kubelet/network/hostport/hostport_manager.go (hunk 2 above)
Comment (thockin) [reply pkg/kubelet/network/hostport/hostport_manager.go]: I think this is better.  It's low cost.

File: pkg/kubelet/network/hostport/hostport_manager.go [hunk 8]
```diff
@@ -177,6 +178,8 @@ func (hm *hostportManager) Remove(id string, podPortMapping *PodPortMapping) (er
 	chainsToRemove := []utiliptables.Chain{}
 	for _, pm := range hostportMappings {
 		chainsToRemove = append(chainsToRemove, getHostportChain(id, pm))
+		// TODO remove this, please refer https://github.com/kubernetes/kubernetes/pull/55153
```
Comment (thockin) [inline pkg/kubelet/network/hostport/hostport_manager.go]: ... after release 1.9

File: pkg/util/iptables/iptables.go (hunk 6 above)
Comment (chenchun) [reply pkg/util/iptables/iptables.go]: It's used by `NewFakeIPTables` to cache builtin chains.

**Comment by thockin** [comment]
This needs an associated issue and a `closes #12345` in the FIRST COMMENT, or it can not be approved.

**Comment by MrHohn** [comment]
Comments are addressed and an associated issue is created.

/lgtm
